Sprint 06

pudramusic opened the pull request.

pudramusic committed:
TestUpload: структура MostPopularMovies имеет свойство items типа [MostPopularMovie], но переменная  mostPopularMovies не имеет такого свойства

pudramusic committed:
TestUpload: структура MostPopularMovies имеет свойство items типа [MostPopularMovie], но переменная  mostPopularMovies не имеет такого свойства

pudramusic committed:
Сдача на ревью Sprint_06

helecta commented on MovieQuiz/Models/GameRecord.swift:
```diff
@@ -8,21 +8,17 @@
-    let total: Int
-    // дата завершения раунда
-    let date: Date
+    let correct: Int // количество  правильных ответов
```
Здорово, что добавляешь комментарии. Если добавлять через /// , а не // , то тогда можно будет быстро прочитать документацию, если зажать кнопку option и навести на класс/свойство/метод.

helecta commented on MovieQuiz/Models/MostPopularMovie.swift:
```diff
@@ -0,0 +1,48 @@
+    private enum Keys: CodingKey {
+        case errorMessage, items
+    }
+    let errorMessage: String
```
Обращу внимание на свойство `errorMessage`. 

Здесь может прийти текст с ошибкой, в таком случае сервис обычно возвращает пустой список фильмов `items`. 
С пустым списком приложение будет выглядеть сломанным. 
Это состояние называют краевым случаем, его стоит обрабатывать как ошибку, например, показывать алерт.
С точки зрения проектирования API сервиса IMDB, 
эту ошибку разработчики должны были вернуть c кодом ошибки клиента ([группа 4хх](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses)). 
Но случается и так, что ответ имеет статус код 200, а тело ответа содержит сообщение об ошибке, как в нашем примере. 
Такой ответ в любом случае необходимо обработать разработчику фронтенда (web, iOS, android).

 **Примеры состояний, когда такое происходит:**

- API key неверный
- API key просрочен
- количество запросов в день превышено (по документации не больше 100)

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -41,21 +34,36 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate,
     
-    func didReceiveNextQuestion(question: QuizQuestion?) {
+    func didLoadDataFromServer() { // метод успешной загрузки данных
+        activityIndicator.isHidden = true // для начала скрываем индикатор загрузки
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped) 
у `activityIndicator`, и использовать его вместо ручного управления свойством `isHidden`. 

Когда ты будешь вызывать `startAnimating` – индикатор будет появляться автоматически, соответственно на `stopAnimating` – скрываться, а свойство `isHidden` больше не понадобится, индикатор все сделает за тебя. 

**Важно:** это свойство настраивают всего раз раз: кодом во `viewDidLoad()` или аналогичной галочкой у элемента на сториборде.
А строки `activityIndicator.isHidden = false(true)` нужно будет убрать

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -41,21 +34,36 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate,
+    func didReceiveNextQuestion(question: QuizQuestion?) {  // метод получения нового вопроса и отображения его на экране. Вопроса может не быть и тогда метод ничего не делает
         guard let question = question else {
             return
         }
```
Ты можешь дополнительно показывать `activityIndicator` между загрузкой вопросов, потому что загрузка вопроса тоже занимает некоторое время из-за получения изображения.
Перед запросом следующего вопроса – показывать индикатор, а в момент получения – выключать анимацию.

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -133,31 +143,40 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate,
+                                  guard let self = self else { return }
+                                  self.currentQuestionIndex = 0
+                                  self.correctAnswers = 0
+                                  self.questionFactory?.requestNextQuestion()
```
Если ошибка произошла из-за того что не было интернета, то данные не загрузились и пытаться запросить следующий вопрос бессмысленно. Тут правильнее снова загружать данные с сервера.

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -133,31 +143,40 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate,
-                self.correctAnswers = 0
-                self.questionFactory.requestNextQuestion()
-                })
+        activityIndicator.isHidden = true // скрываем индикатор
```
Здорово было бы, если ты вынес это в отдельный метод по аналогии с `showLoadingIndicator `, в `hideLoadingIndicator `

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -172,7 +191,7 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate,
             currentQuestionIndex += 1
 
-            self.questionFactory.requestNextQuestion()
+            self.questionFactory?.requestNextQuestion()
```
Тут `self` необязательно использовать, поэтому лучше его убрать.
Подробнее можешь почитать в [код стайле](https://github.com/tutu-ru/swift-style-guide?tab=readme-ov-file#Комментарии)

helecta commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -5,62 +5,108 @@
-    weak var delegate: QuestionFactoryDelegate?
+class QuestionFactory: QuestionFactoryProtocol { //  определяем класс и подписываем на него протокол
+    private let moviesLoader: MoviesLoader // свойство которое реализует протокол MoviesLoader
+    weak var delegate: QuestionFactoryDelegate? // свойство которое реализует протокол QuestionFactoryDelegate
```
Стоит организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/tutu-ru/swift-style-guide?tab=readme-ov-file#Организация-кода)

helecta commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -5,62 +5,108 @@
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL) // загружаем изображение по URL c исправленным размером изображения
+            } catch {
+                print("Failed to load image") // в случае ошибки выводим сообщение
```
Обращу внимание, что игра состоит из вопроса о рейтинге, на который отвечает пользователь, и контекста о фильме в виде постера(картинка). 
Без постера, пользователю будет затруднительно ответить на вопрос о рейтинге фильма верно.

В этом месте, функция не завершает свою работу и идет формировать вопрос без постера. 
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта. 
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

helecta commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -5,62 +5,108 @@
+
+
+    /*
+     private let questions: [QuizQuestion] = [
```
обычно перед сдачей принято удалять закомментированный код и print’ы. Так код проще читать, а дебажная информация не должна попадать в продакшен(т.е print’ы)

pudramusic committed:
Исправление_01_Sprint_06

pudramusic committed:
Исправление_2_Sprint_06
По рекомендации ревьювера мне нужно было вынести индикатор в отдельные две функции - показ индикатора и скрытие.
Мне хочется понять логику:
1) я всегда вызываю функцию показа индикатора во viewDidLoad
2) далее скрывваю этот индикатор между показами вопросов и перед алертой в отдельных функциях
Вопрос:
1) Праввильно ли я сделал что вызываю  функцию скрытия индиктора в функции showAlert и когда прошу делегат вызвать эту функцию, то индикатор будет скрываться?
2) Или все-таки правильно скрывать индикатор в самих алертах?

Так же у меня вопрос по показу индикатора между вопросами - я запускаю индикатор ожидания, но скрыть у меня не получается его. Могли бы подсказать как это сделать?

helecta reviewed (commented):
То как ты сделал, при текущем тз, хорошее решение. Скрывать в самих алертах точно не стоит, т.к лучше чтобы alertPresenter  был максимально независим и не привязан к конкретному контроллеру и ничего не знал об индикатора этого контроллера.
Возможно имело бы смысл вызывать скрытие непосредственно перед вызовом `alertPresenter?.showAlert(alertModel: model)`, т.к необязательно, что для всех алертов в будущем нужно было бы скрывать индикатор.  К примеру, появился бы алерт долгой загрузки, который оповещает пользователя что загрузка долгая и нужно подождать, то в этом случае скрытие делать не нужно.

Но опять повторюсь, при текущем тз твое решение отлично подходит.
